examples/service_worker/src/lib.rs

```diff
@@ -8,112 +8,64 @@
-async fn register_service_worker(
-    msg_sender: Rc<dyn Fn(Option<Msg>)>,
-) -> Result<(), ServiceWorkerError> {
-    let window = web_sys::window().ok_or(ServiceWorkerError::GetWindow)?;
-    let sw_container = window.navigator().service_worker();
-
-    let p = sw_container.register("service-worker.js");
-    let reg = wasm_bindgen_futures::JsFuture::from(p)
-        .await
-        .map_err(ServiceWorkerError::Registration)?;
-
-    let sw_reg: web_sys::ServiceWorkerRegistration = reg.into();
-
-    msg_sender(Some(Msg::SetServiceWorker(sw_reg.clone())));
-    let sw: Option<web_sys::ServiceWorker> = if let Some(x) = sw_reg.installing() {
-        Some(x)
-    } else if let Some(x) = sw_reg.waiting() {
-        Some(x)
-    } else if let Some(x) = sw_reg.active() {
-        Some(x)
-    } else {
-        None
-    };
-
-    if let Some(sw) = sw {
-        if sw.state() == web_sys::ServiceWorkerState::Activated {
-            msg_sender(Some(Msg::ServiceWorkerActivated));
-            msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::Activated)));
-        } else {
-            msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::Waiting)));
-            let c = Closure::wrap(Box::new(move |e: web_sys::Event| {
-                let target: web_sys::EventTarget = e.target().expect("Couldn't get target");
-                let sw_js: JsValue = target.into();
-                let sw: web_sys::ServiceWorker = sw_js.into();
-                let state: web_sys::ServiceWorkerState = sw.state();
-
-                msg_sender(Some(Msg::StatusUpdate(ServiceWorkerStatus::StateChange(
-                    state,
-                ))));
-
-                if state == web_sys::ServiceWorkerState::Activated {
-                    msg_sender(Some(Msg::ServiceWorkerActivated));
-                }
-            }) as Box<dyn Fn(web_sys::Event)>);
-
-            sw.add_event_listener_with_callback("statechange", c.as_ref().unchecked_ref())
-                .map_err(|v| {
-                    sw.remove_event_listener_with_callback(
-                        "statechange",
-                        c.as_ref().unchecked_ref(),
-                    )
-                    .expect("Couldn't remove state change event listener.");
-                    ServiceWorkerError::StateChangeListener(v)
-                })?;
-
-            c.forget();
-        }
-    } else {
-        return Err(ServiceWorkerError::InvalidState);
+    let worker_container = window().navigator().service_worker();
+    let ready = worker_container
+        .ready()
+        .map(JsFuture::from)
+        .expect("get `ready`");
+
+    orders
+        .perform_cmd(async move {
+            Msg::WorkerActivated({
+                ready
+                    .await
+                    .map(web_sys::ServiceWorkerRegistration::from)
+                    .expect("ServiceWorkerRegistration on ready")
+            })
+        })
+        .perform_cmd(async move {
```

Comment:
No, the first future (`ready`) resolves once the worker is registered and active. And the worker is registered in the second future (`register(/service_worker.js..*`). To make sure the `ready` callback/listener is attached, I've written `worker_container.ready().**` before all `perform_cmd` calls and `ready` is executed immediately in `perform_cmd` block.

"Lifetimes" should look like this:
_ready_ Future:
```                                                          
0 |----------------------------------------| x (or to infinite if the worker won't be activated; it can't fail)
```
_register_ Future:
```           
0      |--------------------| y     (y < x)
```
Does it make sense?

Reply:
makes sense 👍